### PR TITLE
Replace RunInDir with RunWithContext

### PR DESCRIPTION
**Pull request by martinscholz83**
Replace `RunInDir` with `RunWIthContext`.

#18553

### CODE REVIEW

**Comment by 6543** [comment]
@martinscholz83 please run `make fmt` and commit the changes

File: integrations/pull_merge_test.go [hunk 1]
```diff
@@ -284,12 +284,13 @@ func TestCantMergeUnrelated(t *testing.T) {
 		assert.NoError(t, err)
 		sha := strings.TrimSpace(stdout.String())
 
-		_, err = git.NewCommand(git.DefaultContext, "update-index", "--add", "--replace", "--cacheinfo", "100644", sha, "somewher-over-the-rainbow").RunInDir(path)
+		err = git.NewCommand(git.DefaultContext, "update-index", "--add", "--replace", "--cacheinfo", "100644", sha, "somewher-over-the-rainbow").RunWithContext(&git.RunContext{Dir: path, Timeout: -1})
 		assert.NoError(t, err)
 
-		treeSha, err := git.NewCommand(git.DefaultContext, "write-tree").RunInDir(path)
+		stdoutTreeSha := new(bytes.Buffer)
```
Comment (lunny) [inline integrations/pull_merge_test.go]: I think `strings.Builder` is better.

File: models/migrations/v134.go [hunk 2]
```diff
@@ -80,7 +81,9 @@ func refixMergeBase(x *xorm.Engine) error {
 
 			gitRefName := fmt.Sprintf("refs/pull/%d/head", pr.Index)
 
-			parentsString, err := git.NewCommand(git.DefaultContext, "rev-list", "--parents", "-n", "1", pr.MergedCommitID).RunInDir(repoPath)
+			stdout := new(bytes.Buffer)
```
Comment (lunny) [inline models/migrations/v134.go]: As above.

File: modules/doctor/mergebase.go [hunk 3]
```diff
@@ -41,20 +42,25 @@ func checkPRMergeBase(ctx context.Context, logger log.Logger, autofix bool) erro
 			repoPath := repo.RepoPath()
 
 			oldMergeBase := pr.MergeBase
+			stdout := new(bytes.Buffer)
```
Comment (lunny) [inline modules/doctor/mergebase.go]: As above.

File: modules/git/commit.go [hunk 4]
```diff
@@ -151,12 +151,13 @@ func AllCommitsCount(ctx context.Context, repoPath string, hidePRRefs bool, file
 		cmd.AddArguments(files...)
 	}
 
-	stdout, err := cmd.RunInDir(repoPath)
+	stdout := new(bytes.Buffer)
```
Comment (lunny) [inline modules/git/commit.go]: As above.

File: modules/git/commit.go [hunk 5]
```diff
@@ -168,12 +169,13 @@ func CommitsCountFiles(ctx context.Context, repoPath string, revision, relpath [
 		cmd.AddArguments(relpath...)
 	}
 
-	stdout, err := cmd.RunInDir(repoPath)
+	stdout := new(bytes.Buffer)
```
Comment (lunny) [inline modules/git/commit.go]: As above.

File: modules/git/remote.go [hunk 6]
```diff
@@ -22,7 +23,9 @@ func GetRemoteAddress(ctx context.Context, repoPath, remoteName string) (*url.UR
 		cmd = NewCommand(ctx, "config", "--get", "remote."+remoteName+".url")
 	}
 
-	result, err := cmd.RunInDir(repoPath)
+	stdout := new(bytes.Buffer)
```
Comment (lunny) [inline modules/git/remote.go]: As above

File: modules/git/repo.go [hunk 7]
```diff
@@ -245,11 +245,12 @@ func Push(ctx context.Context, repoPath string, opts PushOptions) error {
 // GetLatestCommitTime returns time for latest commit in repository (across all branches)
 func GetLatestCommitTime(ctx context.Context, repoPath string) (time.Time, error) {
 	cmd := NewCommand(ctx, "for-each-ref", "--sort=-committerdate", BranchPrefix, "--count", "1", "--format=%(committerdate)")
-	stdout, err := cmd.RunInDir(repoPath)
+	stdout := new(bytes.Buffer)
```
Comment (lunny) [inline modules/git/repo.go]: As above.

File: modules/git/repo.go [hunk 8]
```diff
@@ -262,11 +263,12 @@ type DivergeObject struct {
 func checkDivergence(ctx context.Context, repoPath, baseBranch, targetBranch string) (int, error) {
 	branches := fmt.Sprintf("%s..%s", baseBranch, targetBranch)
 	cmd := NewCommand(ctx, "rev-list", "--count", branches)
-	stdout, err := cmd.RunInDir(repoPath)
+	stdout := new(bytes.Buffer)
```
Comment (lunny) [inline modules/git/repo.go]: As above

File: modules/git/repo_blame.go [hunk 9]
```diff
@@ -13,7 +16,9 @@ func (repo *Repository) FileBlame(revision, path, file string) ([]byte, error) {
 
 // LineBlame returns the latest commit at the given line
 func (repo *Repository) LineBlame(revision, path, file string, line uint) (*Commit, error) {
-	res, err := NewCommand(repo.Ctx, "blame", fmt.Sprintf("-L %d,%d", line, line), "-p", revision, "--", file).RunInDir(path)
+	stdout := new(bytes.Buffer)
```
Comment (lunny) [inline modules/git/repo_blame.go]: As above.

File: modules/git/repo_commit.go [hunk 10]
```diff
@@ -429,9 +434,10 @@ func (repo *Repository) GetCommitsFromIDs(commitIDs []string) []*Commit {
 
 // IsCommitInBranch check if the commit is on the branch
 func (repo *Repository) IsCommitInBranch(commitID, branch string) (r bool, err error) {
-	stdout, err := NewCommand(repo.Ctx, "branch", "--contains", commitID, branch).RunInDir(repo.Path)
+	stdout := new(bytes.Buffer)
```
Comment (lunny) [inline modules/git/repo_commit.go]: As above.

File: modules/git/repo_commit_nogogit.go [hunk 11]
```diff
@@ -18,19 +19,20 @@ import (
 
 // ResolveReference resolves a name to a reference
 func (repo *Repository) ResolveReference(name string) (string, error) {
-	stdout, err := NewCommand(repo.Ctx, "show-ref", "--hash", name).RunInDir(repo.Path)
+	stdout := new(bytes.Buffer)
```
Comment (lunny) [inline modules/git/repo_commit_nogogit.go]: As above

File: modules/git/repo_index.go [hunk 12]
```diff
@@ -118,13 +120,15 @@ func (repo *Repository) RemoveFilesFromIndex(filenames ...string) error {
...
 	cmd := NewCommand(repo.Ctx, "update-index", "--add", "--replace", "--cacheinfo", mode, object.String(), filename)
-	_, err := cmd.RunInDir(repo.Path)
+	err := cmd.RunWithContext(&RunContext{Dir: repo.Path, Timeout: -1})
 	return err
 }
 
 // WriteTree writes the current index as a tree to the object db and returns its hash
 func (repo *Repository) WriteTree() (*Tree, error) {
-	res, err := NewCommand(repo.Ctx, "write-tree").RunInDir(repo.Path)
+	stdout := new(bytes.Buffer)
```
Comment (lunny) [inline modules/git/repo_index.go]: As above

**Comment by lunny** [comment]
I think `strings.Builder` should have better performance than `bytes.Buffer`. Otherwise L-G-T-M .

**Review by lunny — APPROVED** [review]

**Comment by martinscholz83** [comment]
That `repo_tree_gogit.go` build failure I can ignore?

**Comment by lunny** [comment]
> That `repo_tree_gogit.go` build failure I can ignore?

I don't think so.

**Comment by martinscholz83** [comment]
I don't understand why it fails for arm64.

File: modules/git/repo_commit_gogit.go [hunk 13]
```diff
@@ -50,7 +50,9 @@ func (repo *Repository) ConvertToSHA1(commitID string) (SHA1, error) {
 		}
 	}
 
-	actualCommitID, err := NewCommand(repo.Ctx, "rev-parse", "--verify", commitID).RunInDir(repo.Path)
+	stdout := new(strings.Builder)
+	err := NewCommand(repo.Ctx, "rev-parse", "--verify", commitID).RunWithContext(&git.RunContext{Dir: repo.Path, Timeout: -1, Stdout: stdout})
```
Comment (Gusted) [inline modules/git/repo_commit_gogit.go]:
```suggestion
	err := NewCommand(repo.Ctx, "rev-parse", "--verify", commitID).RunWithContext(&RunContext{Dir: repo.Path, Timeout: -1, Stdout: stdout})
```

That should fix the build error.

Comment (martinscholz83) [reply modules/git/repo_commit_gogit.go]: I see

**Comment by lunny** [comment]
Please run `make fmt` and commit again.

**Comment by 6543** [comment]
@martinscholz83 fixed things ;)

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
hmm looks like some stderr is just droped now and we depend on it as err

**Comment by 6543** [comment]
@martinscholz83 can you have a look at it?

use:
`make backend` and  
`make test-sqlite\#TestAPIDeleteTagByName`  
`make test-sqlite\#TestSearchRepo`  
to check

File: services/release/release.go [hunk 14]
```diff
@@ -297,10 +297,11 @@ func DeleteReleaseByID(ctx context.Context, id int64, doer *user_model.User, del
 	}
 
 	if delTag {
-		if stdout, err := git.NewCommand(ctx, "tag", "-d", rel.TagName).
+		stdout := new(strings.Builder)
+		if err := git.NewCommand(ctx, "tag", "-d", rel.TagName).
 			SetDescription(fmt.Sprintf("DeleteReleaseByID (git tag -d): %d", rel.ID)).
-			RunInDir(repo.RepoPath()); err != nil && !strings.Contains(err.Error(), "not found") {
-			log.Error("DeleteReleaseByID (git tag -d): %d in %v Failed:\nStdout: %s\nError: %v", rel.ID, repo, stdout, err)
+			RunWithContext(&git.RunContext{Dir: repo.RepoPath(), Timeout: -1, Stdout: stdout}); err != nil && !strings.Contains(err.Error(), "not found") {
```
Comment (martinscholz83) [inline services/release/release.go]: Her i got
```
Stdout:
Error: exit status 1
```
Is this if statement is valid `err != nil && !strings.Contains(err.Error(), "not found")` ?

Comment (6543) [reply services/release/release.go]: @martinscholz83 take https://github.com/go-gitea/gitea/pull/18978/commits/3305f5a6f52fc29ab58216b121dd621ab5e29604 as example

Comment (martinscholz83) [reply services/release/release.go]: Got it. thx!

**Comment by martinscholz83** [comment]
I don't know if any of my changes does `make test-sqlite\#TestSearchRepo` failing?

File: services/release/release.go [hunk 15]
```diff
@@ -297,10 +297,12 @@ func DeleteReleaseByID(ctx context.Context, id int64, doer *user_model.User, del
...
-		if stdout, err := git.NewCommand(ctx, "tag", "-d", rel.TagName).
+		stdout := new(strings.Builder)
+		stderr := new(strings.Builder)
+		if err := git.NewCommand(ctx, "tag", "-d", rel.TagName).
 			SetDescription(fmt.Sprintf("DeleteReleaseByID (git tag -d): %d", rel.ID)).
-			RunInDir(repo.RepoPath()); err != nil && !strings.Contains(err.Error(), "not found") {
-			log.Error("DeleteReleaseByID (git tag -d): %d in %v Failed:\nStdout: %s\nError: %v", rel.ID, repo, stdout, err)
+			RunWithContext(&git.RunContext{Dir: repo.RepoPath(), Timeout: -1, Stdout: stdout, Stderr: stderr}); err != nil && !strings.Contains(stderr.String(), "not found") {
+			log.Error("DeleteReleaseByID (git tag -d): %d in %v Failed:\nStdout: %s\nError: %v", rel.ID, repo, stdout.String(), stderr.String())
 			return fmt.Errorf("git tag -d: %v", err)
```
Comment (6543) [inline services/release/release.go]:
```suggestion
			return fmt.Errorf("git tag -d: %v", stderr.String())
```

**Comment by 6543** [comment]
.../api/v1/repo/file.go:563:GetContents() [E] GetContentsOrList: exit status 128 

this will need stderr as err too ...

**Comment by lunny** [comment]
CI failed is related.

**Comment by martinscholz83** [comment]
I think this is out of my scope. On my local machine all test run fine. I also did `make fmt` and `make lint-backend`.

**Comment by 6543** [comment]
I'll take a look

**Comment by wxiaoguang** [comment]
Look good to me, 2 cents:
1. Should we introduce a wrapper `RunWithContextString`? It would save a lot of code like `new(strings.Builder)` and `stdout.Reset()
2. Should we remove all `Timeout: -1`, and then treat the default zero value for the default timeout (defaultCommandExecutionTimeout)? If some one want to run a git command for long time, they could set their own Timeout, but not `infinity`, no git command should run infinitely indeed.

**Comment by lunny** [comment]
> Look good to me, 2 cents:
> 
> 1. Should we introduce a wrapper `RunWithContextString`? It would save a lot of code like `new(strings.Builder)` and `stdout.Reset()`
> 2. Should we remove all `Timeout: -1`, and then treat the default zero value as the default timeout (defaultCommandExecutionTimeout)? And currently the zero value doesn't have clear meaning, when it is used by `context.WithTimeout`, zero means timeout immediately, which is meaningless.

In fact, we need a no timeout option.

**Comment by wxiaoguang** [comment]
> > Look good to me, 2 cents:
> > 
> > 1. Should we introduce a wrapper `RunWithContextString`? It would save a lot of code like `new(strings.Builder)` and `stdout.Reset()`
> > 2. Should we remove all `Timeout: -1`, and then treat the default zero value as the default timeout (defaultCommandExecutionTimeout)? And currently the zero value doesn't have clear meaning, when it is used by `context.WithTimeout`, zero means timeout immediately, which is meaningless.
> 
> In fact, we need a no timeout option.

If some one want to run a git command for long time, they could set their own Timeout (100 or 99999 years if they like), but not infinity (or no timeout).

**Comment by wxiaoguang** [comment]
For this PR, it's so weird. CI always fails in `test-sqlite` (while others all succeed), the error is always `GetContentsOrList/GetCommit exit status 128`, I have tried to restart the CI many times, all the same. I tried it locally and it seems that the test can succeed ... maybe need some more investigation.

**Comment by lunny** [comment]
Please resolve the conflicts.

**Comment by martinscholz83** [comment]
One more thing for my understanding. I found a lot places like [this](https://github.com/martinscholz83/gitea/blob/feature/RunInPipline/services/migrations/gitea_uploader.go#L555) where still the ctx error is logged instead of the output from `stderr`. So my question, do these places also need to be changed to log `stderr` instead of `err`?

**Comment by 6543** [comment]
@martinscholz83 a general roule: if err is just needed to check if it fails we dont need the stderr, if it is used in logs etc .. stderr is needed

and I agree to add a wrapper called `RunWithContextString` that return **(stdout, stderr, err)** - that way we dont need mouch boilerplate

**Comment by zeripath** [comment]
Hmm... although this fails in different places each time I think it's been consistently failing too many times and we need to double check that there isn't a real issue here.

**Comment by wxiaoguang** [comment]
CI failure resolved. Then I will remove the `Timeout: -1` and use `RunWithContextString` for some cases.

And, I believe there are some bugs during the early refactoring, mainly caused by using `err.Error()` wrongly instead of `stderr`

I would try my best to clean them, but I can not promise I can get all them cleaned at one time.

File: modules/git/command.go [hunk 16]
```diff
@@ -203,6 +204,19 @@ func (c *Command) RunWithContext(rc *RunContext) error {
 	return ctx.Err()
 }
 
+// RunWithContextString run the command with context and returns stdout/stderr as string
+func (c *Command) RunWithContextString(rc *RunContext) (stdout, stderr string, err error) {
+	if rc.Stdout != nil || rc.Stderr != nil {
+		panic("stdout and stderr field must be nil when using RunWithContextString")
```
Comment (lunny) [inline modules/git/command.go]: I don't think we need a panic here.

**Review by lunny — CHANGES_REQUESTED** [review]

File: modules/git/repo_commit_gogit.go [hunk 17]
```diff
@@ -50,16 +50,16 @@ func (repo *Repository) ConvertToSHA1(commitID string) (SHA1, error) {
 		}
 	}
 
-	actualCommitID, err := NewCommand(repo.Ctx, "rev-parse", "--verify", commitID).RunInDir(repo.Path)
+	stdout, stderr, err := NewCommand(repo.Ctx, "rev-parse", "--verify", commitID).RunWithContextString(&RunContext{Dir: repo.Path, Timeout: -1})
```
Comment (lunny) [inline modules/git/repo_commit_gogit.go]: `RunWithContextString` is only invoked once here. It's really necessary?

File: modules/git/command.go (hunk 16 above)
Comment (wxiaoguang) [reply modules/git/command.go]: It does need.

**Comment by wxiaoguang** [comment]
This PR does wrong from begnning ..... replacing `RunInDir` by `RunWithContext` is incorrect, because the error returned by `RunInDir` contains stderr, while `RunWithContext` doesn't. Such replacing causes many error checks doesn't work correctly.

File: modules/git/repo_commit_gogit.go (hunk 17 above)
Comment (wxiaoguang) [reply modules/git/repo_commit_gogit.go]: See my comment

File: modules/git/command.go [hunk 18]
```diff
@@ -239,30 +239,12 @@ func (c *Command) RunInDirTimeoutEnv(env []string, timeout time.Duration, dir st
...
-
 // RunInDirBytes executes the command in given directory
 // and returns stdout in []byte and error (combined with stderr).
 func (c *Command) RunInDirBytes(dir string) ([]byte, error) {
 	return c.RunInDirTimeout(-1, dir)
 }
 
-// RunInDir executes the command in given directory
-// and returns stdout in string and error (combined with stderr).
-func (c *Command) RunInDir(dir string) (string, error) {
```
Comment (wxiaoguang) [inline modules/git/command.go:262]: The key point is here, it returns "stdout in string and error (**combined with stderr**)"

So the `RunInXxx` can not be replaced by `RunWithContext` directly

**Comment by wxiaoguang** [comment]
Hi @martinscholz83 , thank you for your contribution and give us the hint about the refactoring. Now we can close #18553 and we have a clear git command module in Gitea!  🎉

Although this PR is not merged, it really help us to know how to continue about the refactoring. Appreciate your work.

Now we can safely close this PR.

**Pull request closed without merging**